src/lib/structures/uid.ml

```diff
@@ -0,0 +1,98 @@
+(**************************************************************************)
+(*                                                                        *)
+(*     Alt-Ergo: The SMT Solver For Software Verification                 *)
+(*     Copyright (C) 2013-2024 --- OCamlPro SAS                           *)
+(*                                                                        *)
+(*     This file is distributed under the terms of OCamlPro               *)
+(*     Non-Commercial Purpose License, version 1.                         *)
+(*                                                                        *)
+(*     As an exception, Alt-Ergo Club members at the Gold level can       *)
+(*     use this file under the terms of the Apache Software License       *)
+(*     version 2.0.                                                       *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     The Alt-Ergo theorem prover                                        *)
+(*                                                                        *)
+(*     Sylvain Conchon, Evelyne Contejean, Francois Bobot                 *)
+(*     Mohamed Iguernelala, Stephane Lescuyer, Alain Mebsout              *)
+(*                                                                        *)
+(*     CNRS - INRIA - Universite Paris Sud                                *)
+(*                                                                        *)
+(*     Until 2013, some parts of this code were released under            *)
+(*     the Apache Software License version 2.0.                           *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module DStd = Dolmen.Std
+module DE = DStd.Expr
+
+type t =
+  | Fake of Hstring.t
+  | Unique of { name : Hstring.t; index : int }
```

Comment:
Of course I tried to use the `Dolmen identifier` directly but I got a regression in our non-regression test suite. I should investigate.

Reply:
Ok, I tried again with the original index produced by Dolmen. Now I got five regressions and it seems that Alt-Ergo ignores the timeout for them (`make runtest` never returns). Actually, if I run `dune exec -- alt-ergo tests/adts/match_in_terms.ae`, I cannot kill the processus with `CTRL+C`. I have to type `pkill alt-ergo`.

Reply:
I suspect an issue with `DE.Id.compare` or `DE.Id.equal`.

Reply:
Uhm they seems to be reasonable: 
```ocaml 
  let compare v v' = compare v.index v'.index

  let equal v v' = compare v v' = 0
```
very strange ...

Reply:
Thanks to Vincent, I got the bug with `gdb`. The issue comes from the fact that `Dolmen` identifiers can be infinite value and the polymorphic equality loops infinitely on it. I wrote another PR to remove as much as possible polymorphic comparison in lists.

Reply:
Yeah, dolmen identifiers should never be compared using the polymorphic identity: besides being slow (compared to the comparison defined by dolmen), it may loop forever (because the `Type` identifiers loops on itself), or may fail at runtime (because ids contain tags that can contain arbitrary data such as functions).